src/gui/qgsfieldmappingmodel.cpp

```diff
@@ -0,0 +1,434 @@
+          break;
+        }
+      }
+      emit dataChanged( index, index );
+    }
+  }
+  return true;
+}
+
+QgsFieldConstraints::Constraints QgsFieldMappingModel::fieldConstraints( const QgsField &field ) const
+{
+  QgsFieldConstraints::Constraints constraints;
+
+  const QgsFieldConstraints fieldConstraints { field.constraints() };
+
+  if ( fieldConstraints.constraints() & QgsFieldConstraints::ConstraintNotNull &&
+       fieldConstraints.constraintStrength( QgsFieldConstraints::ConstraintNotNull ) & QgsFieldConstraints::ConstraintStrengthHard )
+    constraints.setFlag( QgsFieldConstraints::ConstraintNotNull );
+
+  if ( fieldConstraints.constraints() & QgsFieldConstraints::ConstraintUnique &&
+       fieldConstraints.constraintStrength( QgsFieldConstraints::ConstraintUnique ) & QgsFieldConstraints::ConstraintStrengthHard )
+    constraints.setFlag( QgsFieldConstraints::ConstraintUnique );
+
+  if ( fieldConstraints.constraints() & QgsFieldConstraints::ConstraintExpression &&
+       fieldConstraints.constraintStrength( QgsFieldConstraints::ConstraintExpression ) & QgsFieldConstraints::ConstraintStrengthHard )
+    constraints.setFlag( QgsFieldConstraints::ConstraintExpression );
+
+  return constraints;
+}
+
+bool QgsFieldMappingModel::moveUpOrDown( const QModelIndex &index, bool up )
+{
+  if ( ! index.isValid() && index.model() == this )
+    return false;
+
+  // Always swap down
+  const int row { up ? index.row() - 1 : index.row() };
+  // Range checking
+  if ( row < 0 || row + 1 >= rowCount( QModelIndex() ) )
+  {
+    return false;
+  }
+  beginMoveRows( QModelIndex( ), row, row, QModelIndex(), row + 2 );
+#if QT_VERSION < QT_VERSION_CHECK(5, 13, 0)
+  mMapping.swap( row, row + 1 );
+#else
+  mMapping.swapItemsAt( row, row + 1 );
+#endif
+  endMoveRows();
+  return true;
+}
+
+QString QgsFieldMappingModel::bestMatchforField( const QgsFieldMappingModel::Field &f, QStringList &excludedFieldNames )
+{
+  // Search for fields in the source
+  // 1. match by name
+  for ( const auto &sf : qgis::as_const( mSourceFields ) )
+  {
+    if ( sf.name() == f.field.name() )
+    {
+      return QgsExpression::quotedColumnRef( sf.name() );
+    }
+  }
+  // 2. match by type
+  for ( const auto &sf : qgis::as_const( mSourceFields ) )
+  {
+    if ( excludedFieldNames.contains( sf.name() ) || sf.type() != f.field.type() )
+      continue;
+    excludedFieldNames.push_back( sf.name() );
+    return QgsExpression::quotedColumnRef( sf.name() );
+  }
+  return QString();
+}
+
+void QgsFieldMappingModel::setSourceFields( const QgsFields &sourceFields )
+{
+  mSourceFields = sourceFields;
+  QStringList usedFields;
+  beginResetModel();
+  for ( const Field &f : qgis::as_const( mMapping ) )
+  {
+    if ( f.expression.isField() )
+    {
+      usedFields.push_back( f.expression.expression().mid( 1, f.expression.expression().length() - 2 ) );
+    }
+  }
+  // not const on purpose
+  for ( Field &f : mMapping )
```

Comment:
I don't think you're allowed to edit a qt container while iterating over it this way. You'd need to do the full `for ( auto it = mMapping.begin(), .... ) ` approach.